### ISSUE TITLE
lib/warnings.t: blocker to strict-by-default

### ISSUE THREAD

**Issue by jkeenan** [atoomic/perl#235]
As of commit ac6b7ccf73 in the alpha-dev-02-strict branch on Aug 18 2020, the file cited is failing.  Here is output from `cd t; ./perl harness -v <file>; cd -`:
```
# ../lib/warnings.t

# Over 6000 lines of output, trimmed here to:

 head -n 20 err
PROG: -w

# warnable code, warnings enabled via command line switch
$c =+ 3 ;
EXPECTED:
Reversed += operator at - line 3.
Name "main::c" used only once: possible typo at - line 3.
GOT:
Reversed += operator at - line 3.
Global symbol "$c" requires explicit package name (did you forget to declare "my $c"?) at - line 3.
Execution of - aborted due to compilation errors.
# Failed test 2 - at lib/warnings/1global line 11
PROG: 
#! perl -w
# warnable code, warnings enabled via #! line
$c =+ 3 ;
EXPECTED:
Reversed += operator at - line 3.
Name "main::c" used only once: possible typo at - line 3.
GOT:
```
```
$ tail -n 20 err
	(less 3 skipped subtests: 707 okay)

Test Summary Report
-------------------
../lib/warnings.t (Wstat: 0 Tests: 917 Failed: 207)
  Failed tests:  2-7, 35, 223-224, 236, 246-249, 257-258
                262, 264, 272, 276-277, 279, 281, 284-288
                298, 304, 310, 339, 354, 358-359, 362, 364
                366, 369, 378, 396-399, 404-416, 431-436
                442, 444-445, 447, 456-457, 459-471, 482
                484-485, 488, 492, 494-495, 504-506, 508
                520-521, 524-526, 528, 545, 567, 575, 577-602
                610-611, 616, 621, 624-625, 628, 633, 637-638
                640, 643-645, 656-657, 661, 663, 668, 673
                683-687, 692, 705-708, 715-716, 718-721
                723-729, 758-759, 764, 769-773, 781-789
                813-816, 819, 822-824, 829, 831, 847-848
                865, 867, 871, 874-875, 906-907
Files=1, Tests=917,  7 wallclock secs ( 0.28 usr  0.02 sys +  4.68 cusr  1.58 csys =  6.56 CPU)
Result: FAIL
```

@atoomic @toddr 

Thank you very much.
Jim Keenan

**Comment by atoomic** [comment]
this is fixed by 8ea8acba

**Comment by jkeenan** [comment]
> this is fixed by [8ea8acb](https://github.com/atoomic/perl/commit/8ea8acbae7da1948fc4ef8f7a9bb70e30e1d49da)

Thanks for this effort.  It led to a dramatic reduction in the lines of output from `make test`.

I see two problematic cases.

**Case 1:** Changed expectation of what test should output

There is one instance is which you've changed the output which the test `EXPECT`s.  Can you say why that was necessary?

```
diff --git a/t/lib/warnings/1global b/t/lib/warnings/1global
index 51acd17850..e046d0ec9d 100644
--- a/t/lib/warnings/1global
+++ b/t/lib/warnings/1global
@@ -13,11 +13,19 @@ EXPECT
 $c =+ 3 ;
 EXPECT
 Reversed += operator at - line 3.
+Global symbol "$c" requires explicit package name (did you forget to declare "my $c"?) at - line 3.
+Execution of - aborted due to compilation errors.
+########
+-w
+# warnable code, warnings enabled via command line switch
+$main::c =+ 3 ;
+EXPECT
+Reversed += operator at - line 3.
 Name "main::c" used only once: possible typo at - line 3.
```

**Case 2**: Relaxation of strictures should be made more precise

There are at 73 cases where you got a test to pass simply by relaxing strictures entirely.
```
$ ack '^\+.*?no\s+strict;$' out
+no warnings 'reserved'; no strict;
+use warnings 'uninitialized'; no strict;
+use warnings 'uninitialized'; no strict;
+use warnings 'uninitialized'; no strict;
+use warnings 'uninitialized'; no strict;
+use warnings 'uninitialized'; no strict;
+use warnings 'uninitialized'; no strict;
+use warnings; no strict;
+use warnings; no strict;
+use warnings; no strict;
+use warnings; no strict;
+use warnings; no strict;
+use warnings; no strict;
+use warnings; no strict;
+use warnings; no strict;
+use warnings; no strict;
+use warnings; no strict;
+use warnings; no strict;
+use warnings; no strict;
+use warnings 'void' ; no strict;
+use utf8; no strict;
+use warnings 'misc' ; use utf8; no strict;
+use warnings 'parenthesis' ; no strict;
+use warnings 'parenthesis' ; no strict;
+use warnings 'parenthesis' ; no strict;
+use warnings 'bareword' ; no strict;
+use warnings 'misc' ; no strict;
+use warnings 'misc' ; no strict;
+use warnings 'misc' ; no strict;
+use warnings 'syntax' ; no strict;
+use warnings 'syntax' ; no strict;
+use warnings 'precedence'; no strict;
+use integer; no strict;
+use warnings 'syntax'; no strict;
+use warnings; no strict;
+use warnings; no strict;
+no strict;
+no strict;
+use warnings 'once'; no strict;
+no strict;
+no warnings 'once' ; no strict;
+use open qw( :utf8 :std ); no strict;
+use utf8; no strict;
+use utf8; no strict;
+use utf8; no strict;
+use utf8; no strict;
+use utf8; no strict;
+use utf8; no strict;
+use utf8; no strict;
+use utf8; no strict;
+use utf8; no strict;
+use utf8; no strict;
+use utf8; no strict;
+use utf8; no strict;
+use warnings 'once'; no strict;
+use warnings 'once'; no strict;
+no strict;
+use warnings 'uninitialized'; no strict;
+use warnings 'closed'; no strict;
+use warnings 'uninitialized' ; no strict;
+use warnings 'uninitialized' ; no strict;
+use Config; no strict;
+use Config; no strict;
+no strict;
+use warnings 'ambiguous' ; no strict;
+use warnings 'syntax' ; no strict;
+no strict;
+no strict;
+use utf8; no strict;
+use utf8; no strict;
+use utf8; no strict;
+use utf8; no strict;
+no warnings; no strict;
```
(This probably applies to additional cases where `no strict;` was inserted into the middle of a line rather than appended to the end of a line as in the `ack` pattern above.)

Would it be possible to be more precise with respect to which strictures we are relaxing in each instance?

For example, take this section of `t/lib/warnings/9uninit`:
```
 305 ########
 306 use warnings 'uninitialized'; no strict;
 307 my ($m1);
 308 our ($g1);
 309 
 310 close $m1;  # exercises rv2gv
 311 close $g1;  # exercises rv2gv
 312 close undef;    # exercises rv2gv
 313 EXPECT
 314 Use of uninitialized value $m1 in ref-to-glob cast at - line 5.
 315 Use of uninitialized value $g1 in ref-to-glob cast at - line 6.
 316 Use of uninitialized value in ref-to-glob cast at - line 7.
 317 ########
```
Experimentation showed that the only type of stricture which needed to be relaxed here in order to get the test to pass was `refs`:
```
diff --git a/t/lib/warnings/9uninit b/t/lib/warnings/9uninit
index 1f0efd4b0b..f93d25f722 100644
--- a/t/lib/warnings/9uninit
+++ b/t/lib/warnings/9uninit
@@ -303,7 +303,7 @@ Use of uninitialized value in print at - line 7.
 Use of uninitialized value $m2 in print at - line 7.
 Use of uninitialized value in print at - line 8.
 ########
-use warnings 'uninitialized'; no strict;
+use warnings 'uninitialized'; no strict 'refs';
 my ($m1);
 our ($g1);
```
I concede that relaxing all strictures is the easiest approach.  But as I worked through the rest of the test suite I started to realize that that was overkill.  Eventually I was able to recognize which kind of stricture needed relaxation in which particular instance.  I think that makes the test code more maintainable going forward.

Thank you very much.
Jim Keenan

**Comment by jkeenan** [comment]
On 8/19/20 9:22 PM, James E Keenan wrote:
>     this is fixed by 8ea8acb
>     <https://github.com/atoomic/perl/commit/8ea8acbae7da1948fc4ef8f7a9bb70e30e1d49da>
> 

First version of this comment had a copy-and-paste error which 
duplicated most of the text.  I edited the comment in the GUI to remove 
the duplication.  The better version starts here:

> 
> Thanks for this effort. It led to a dramatic reduction in the lines of 
> output from |make test|.
> 
> I see two problematic cases.
> 
> *Case 1:* Changed expectation of what test should output
> 
> There is one instance is which you've changed the output which the test 
> |EXPECT|s. Can you say why that was necessary?

... and so forth.

jimk

**Comment by atoomic** [comment]
The problem is that enabling strict for some of these subtests will change the whole purpose of the test itself...
as with strict enabled this will trigger a failure at compile time and we would not guarantee the expected behavior we are trying to test.

I avoided to add `use strict`, as much as possible.I might probably oversight a few of them.
But removing `use strict` to these tests is definitively not the right choice IMO.

Note that the `EXPECT` part changes are mainly reflecting the line number changes due to the addition of additional code like variable declaration or other.

**Comment by jkeenan** [comment]
> The problem is that enabling strict for some of these subtests will change the whole purpose of the test itself...
> as with strict enabled this will trigger a failure at compile time and we would not guarantee the expected behavior we are trying to test.

I think there's some confusion here.  By virtue of being in this Perl 7 branch, these tests **already** have 'strict' enabled.  So if we want to revise the test to reflect that environmental change but retain the same `EXPECT`, we have to **turn off** strictures.  But we should do that judiciously, *i.e.,* by turning off only the strictures needed to get the unit test to pass.

> [snip]
> 
> Note that the `EXPECT` part changes are mainly reflecting the line number changes due to the addition of additional code like variable declaration or other.

Yes, I encountered that same phenomenon in other test files.

Thank you very much.
Jim Keenan

**Comment by atoomic** [comment]
I meant I avoided to add `no strict;` as much as possible but in some cases it was necessary to preserve the test intent

**Comment by jkeenan** [comment]
The problems described in Case 2 above have been substantially addressed by commit c20b1c45f0.  Closing ticket.
